src/librustc/middle/stability.rs

```diff
@@ -764,13 +764,15 @@ impl<'a, 'tcx> Visitor<'tcx> for Checker<'a, 'tcx> {
                     emit_feature_err(&self.tcx.sess.parse_sess,
                                      "untagged_unions", item.span, GateIssue::Language,
                                      "unions with `Drop` implementations are unstable");
-                } else {
-                    let param_env = self.tcx.param_env(def_id);
-                    if !param_env.can_type_implement_copy(self.tcx, ty).is_ok() {
-                        emit_feature_err(&self.tcx.sess.parse_sess,
-                                         "untagged_unions", item.span, GateIssue::Language,
-                                         "unions with non-`Copy` fields are unstable");
-                    }
+                }
+
+                let param_env = self.tcx.param_env(def_id);
+                let field_needs_drop = adt_def.non_enum_variant().fields.iter().any(|field| {
+                        self.tcx.type_of(field.did).needs_drop(self.tcx, param_env)
+                });
+                if field_needs_drop {
+                    self.tcx.sess.span_err(item.span, "unions may not contain fields \
+                                                       that need dropping");
```

Comment:
Note that this is still under `!self.tcx.features().untagged_unions`, so the error won't get triggered with `#![feature(untagged_unions)]`, which is likely not what we want.

Also, it should *probably* not be in this file if it's not tied to stability checks.

cc @nikomatsakis Do you have any suggestions for where to place this code?

Reply:
This is some form of well-formedness check of a type declaration. Where do we usually do these? Things like "recursive occurrences must be below a pointer indirection", "only the last field of a struct may be unsized", ...